datasets/cbt/cbt.py

```diff
@@ -0,0 +1,335 @@
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir, paths[self.config.name]["train"]),
+                },
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.TEST,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir, paths[self.config.name]["test"]),
+                },
+            ),
+            datasets.SplitGenerator(
+                name=datasets.Split.VALIDATION,
+                # These kwargs will be passed to _generate_examples
+                gen_kwargs={
+                    "filepath": os.path.join(data_dir, paths[self.config.name]["valid"]),
+                },
+            ),
+        ]
+
+    def _generate_examples(
+        self, filepath  # method parameters are unpacked from `gen_kwargs` as given in `_split_generators`
+    ):
+        """ Yields examples as (key, example) tuples. """
+        # This method handles input defined in _split_generators to yield (key, example) tuples from the dataset.
+        # The `key` is here for legacy reason (tfds) and is not important in itself.
+
+        if self.config.name != "raw":
+            with open(filepath, encoding="utf-8") as f:
+                data = f.readlines()
+                sentences = []
+                example_idx = 0
+                for idx, line in enumerate(data):
+                    if line.strip() == "":
+                        continue
+
+                    elif line.split()[0] == "21":
+                        splitline = line.split("\t")  # question, answer options are tab separated
+                        question = splitline[0]
+                        answer = splitline[1]
+                        options = splitline[-1]
+                        question = question[2:].strip()  # The first two indices contain `21`.
+                        answer = answer.strip()
+                        options = options.strip().split("|")
+                        yield example_idx, {
+                            "sentences": sentences,
+                            "question": question,
+                            "options": options,
+                            "answer": answer,
+                        }
+
+                        sentences = []
+                        example_idx += 1
+                    else:
+                        if len(line.split()[0]) == 1:
+                            sentences.append(line[1:].strip())
+                        else:
+                            sentences.append(line[2:].strip())
+                            # Text might contain double spaces.
+        else:
+            with open(filepath, encoding="utf=8") as f:
+                data = f.read()
```

Comment:
same here

Reply:
Here I need to search over the entire data for `_BOOK_TITLE_`, it's not possible to do this without reading entire data.

The alternative is to look for lines and check the first few characters for `_BOOK_TITLE_`. So, I'll do that.